### ISSUE TITLE
say vs print vs show

### ISSUE THREAD

**Issue by trans** [evincarofautumn/kitten#204]
Could you explain the the difference between `say`, `print` and `show`?

**Comment by Camto** [comment]
`say` outputs a string with a newline appended. `print` doesn't add a newline. `show` takes something and returns it as a string without outputting it.

**Comment by sullyj3** [comment]
The name "show" is probably derived from Haskell, where it has the same functionality of just returning a string representation.

**Comment by evincarofautumn** [comment]
Yeah, `show` was borrowed from Haskell. (Incidentally, `say` is from Perl.) If anyone has other name ideas, I’m open to suggestions. This area isn’t very thought out.

Ideally I think `show` should be split into two traits: one for debug output (`show`, `debug`, `repr`, `ugly`?), and one (or more) for pretty-printing (`pretty`, `human`, `show`?), which is meant for producing nice human-readable output in a user interface. The debug trait should be derived automatically, produce a valid Kitten expression, show the full structure of a value, and have an inverse trait (`read`, `parse`?) for debug input. The pretty-printing trait might return plain text, or some other value representing a graphical structure that can be rendered in a terminal (for the interactive mode) or otherwise (hypothetically, something like Jupyter/IPython notebooks).

**Comment by brendanzab** [comment]
Rust has `fmt::Debug` and `fmt::Display` if you're looking for prior art. I'd probably recommend `print` and `println` for printing without/with newlines.

**Comment by brendanzab** [comment]
`ugly` is kinda cute though!

**Comment by trans** [comment]
`ln` could just be a separate word for sending a newline to standard out. Then `print ln` is equivalent to `say` (and `println` in other languages). Also has the advantage of simplifying `"" say` just to print a newline.

If `show` is to get a canonical string representation, how does that generally fit into type representation conversion in general. eg, how does one take a string and turn it into an integer? Almost seems like an operator makes sense, `~> String`, or something.

Maybe `log` for debug trait?

For pretty-print trait, maybe `pp` like in Ruby.

**Comment by sullyj3** [comment]
That's very cute.
